src/comments.ts

```diff
@@ -0,0 +1,297 @@
+import vscode from 'vscode'
+import { repoInfo } from './git'
+import { gql, mutateGraphQL, queryGraphQL } from './graphql'
+import { log } from './log'
+
+export function activateComments(context: vscode.ExtensionContext): void {
+    if (!vscode.workspace.getConfiguration('sourcegraph').get<string>('discussionsPreviewEnabled')) {
+        return
+    }
+    log.appendLine('Discussions preview is enabled')
+    const commentProvider = new CommentProvider()
+    context.subscriptions.push(vscode.workspace.registerDocumentCommentProvider(commentProvider))
+}
+
+class CommentProvider implements vscode.DocumentCommentProvider {
+    public async provideDocumentComments(
+        document: vscode.TextDocument,
+        token: vscode.CancellationToken
+    ): Promise<vscode.CommentInfo> {
+        let threads: vscode.CommentThread[] = []
+        try {
+            threads = await provideDocumentComments(document)
+        } catch (e) {
+            log.appendLine(`provideDocumentComments ${e}`)
+        }
+
+        const lastLine = document.lineCount - 1
+        const commentingRanges = [new vscode.Range(0, 0, lastLine, document.lineAt(lastLine).range.end.character)]
+
+        return { threads, commentingRanges }
+    }
+
+    public async createNewCommentThread(
+        document: vscode.TextDocument,
+        range: vscode.Range,
+        text: string,
+        token: vscode.CancellationToken
+    ): Promise<vscode.CommentThread> {
+        return await createNewCommentThread(document, range, text)
+    }
+
+    public async replyToCommentThread(
+        document: vscode.TextDocument,
+        range: vscode.Range,
+        commentThread: vscode.CommentThread,
+        text: string,
+        token: vscode.CancellationToken
+    ): Promise<vscode.CommentThread> {
+        return await replyToCommentThread(document, range, commentThread, text)
+    }
+
+    private didChangeCommentThreads = new vscode.EventEmitter<vscode.CommentThreadChangedEvent>()
+
+    public onDidChangeCommentThreads = this.didChangeCommentThreads.event
+}
+
+async function provideDocumentComments(document: vscode.TextDocument): Promise<vscode.CommentThread[]> {
+    log.appendLine(`provideDocumentComments ${document.uri}`)
+    const [remoteUrl, branch, path] = await repoInfo(document.fileName)
+    if (remoteUrl === '') {
```

Comment:
I would recommend to check for falsiness because this would likely get refactored to return `undefined` at some point instead of empty strings